src/v/kafka/server/handlers/fetch.h

```diff
@@ -328,11 +323,20 @@ struct shard_fetch {
 };
 
 struct fetch_plan {
-    explicit fetch_plan(size_t shards)
-      : fetches_per_shard(shards) {}
+    explicit fetch_plan(
+      size_t shards,
+      op_context::latency_point start_time = op_context::latency_clock::now())
+      : fetches_per_shard(shards, shard_fetch(start_time)) {}
 
     std::vector<shard_fetch> fetches_per_shard;
 
+    void reserve_from_partition_count(size_t count) {
+        size_t per_partition = count * 3 / (fetches_per_shard.size() * 2);
```

Comment:
Basically reserve 1.5x the "average" size we'd expect if the involved partitions were evenly distributed among shards. Nothing super scientific about this: we want to be more than 1.0 since that will result in reallocations right at the end (the worst kind!) for many shards if there even a slight imabalance, which seems likely, but we don't want to go overboard either.

Reply:
we need to somehow reflect that in the fetch estimator [here](https://github.com/redpanda-data/redpanda/pull/10371/commits/f38586087b4f9426906469c8f4a3164850c4e690) when both PRs land

Reply:
@dlex - I guess yeah. That said, it's not clear how much this increases memory use by. In general `reserve` decreases memory use since the memory region is exactly sized. That's not the case here due to the 3/2 factor (and the unknown distribution among shards) but even before the vector has an average over-allocation of 1.5x when it is grown via push_back. 

Were we accounting for that?